graphstorm-processing/graphstorm_processing/distributed_executor.py

```diff
@@ -0,0 +1,342 @@
+    ):
+        self.local_config_path = executor_config.local_config_path
+        self.local_output_path = executor_config.local_output_path
+        self.input_prefix = executor_config.input_prefix
+        self.output_prefix = executor_config.output_prefix
+        self.num_output_files = executor_config.num_output_files
+        self.config_filename = executor_config.config_filename
+        self.filesystem_type = executor_config.filesystem_type
+        self.sm_execution = executor_config.sm_execution
+        self.add_reverse_edges = executor_config.add_reverse_edges
+
+        graph_conf = os.path.join(self.local_config_path, self.config_filename)
+        with open(graph_conf, "r", encoding="utf-8") as f:
+            dataset_config_dict: Dict[str, Any] = json.load(f)
+
+        if "version" in dataset_config_dict:
+            self.config_version = dataset_config_dict["version"]
+            self.graph_config_dict = dataset_config_dict["graph"]
+        else:
+            # TODO: Change once GConstruct adds a version to their config spec
+            self.config_version = "gconstruct"
+            converter = GConstructConfigConverter()
+            self.graph_config_dict = converter.convert_to_gsprocessing(dataset_config_dict)["graph"]
+
+        # Create the Spark session for execution
+        self.spark = spark_utils.create_spark_session(self.sm_execution, self.filesystem_type)
+
+    def run(self) -> None:
+        """
+        Executes the Spark processing job.
+        """
+        print("Performing data processing with PySpark...")
+        data_configs = create_config_objects(self.graph_config_dict)
+
+        t0 = time.time()
+        print("Constructing DGLGraph for Heterogeneous Graph")
+        # Prefer explicit arguments for clarity
+        loader = DistHeterogeneousGraphLoader(
+            spark=self.spark,
+            local_input_path=self.local_config_path,
+            local_output_path=self.local_output_path,
+            data_configs=data_configs,
+            input_prefix=self.input_prefix,
+            output_prefix=self.output_prefix,
+            num_output_files=self.num_output_files,
+            add_reverse_edges=self.add_reverse_edges,
+            enable_assertions=False,
+        )
+        loader.load()
+
+        # This is used to upload the output JSON files to S3 on local runs,
+        # since we can't rely on SageMaker to do it
+        if not self.sm_execution and self.filesystem_type == "s3":
+            bucket = self.output_prefix.split("/")[2]
+            s3_prefix = self.output_prefix.split("/", 3)[3]
+            s3 = boto3.resource("s3")
+
+            output_files = os.listdir(loader.output_path)
+            for output_file in output_files:
+                s3.meta.client.upload_file(
+                    f"{os.path.join(loader.output_path, output_file)}",
+                    bucket,
+                    f"{s3_prefix}/{output_file}",
+                )
```

Comment:
Shall we remove the files under "/opt/ml/processing/output" ?
Will SageMaker upload those files automatically?

Reply:
It will upload them but that's by design.

The only things that are created under there are the job's JSON outputs: 
* Metadata JSON used by re-partitioning/dist-partitioning
* Performance counters JSON
* Launch arguments JSON.

So there's minimal overhead in uploading these.